packages/react-storage/src/components/StorageBrowser/views/LocationActionView/CopyView/useCopyView.ts

```diff
@@ -1,5 +1,95 @@
+import { useState } from 'react';
+
+import { isFunction } from '@aws-amplify/ui';
+
+import { copyHandler } from '../../../actions/handlers';
+import { useProcessTasks } from '../../../tasks';
+import { useGetActionInput } from '../../../providers/configuration';
+import { useStore } from '../../../providers/store';
+import { getActionViewDisabledButtons } from '../utils';
+import { getTaskCounts } from '../../../controls/getTaskCounts';
+import { getDestinationListFullPrefix } from '../utils/getDestinationPickerDataTable';
 import { CopyViewState } from './types';
 
-export function useCopyView(): CopyViewState {
-  throw new Error('useCopyView is not implemented');
-}
+export const useCopyView = ({
+  onExit: _onExit,
+}: {
+  onExit?: () => void;
+}): CopyViewState => {
+  const [
+    {
+      location,
+      locationItems: { fileDataItems },
+    },
+    dispatchStoreAction,
+  ] = useStore();
+  const { key, current } = location;
+
+  const getInput = useGetActionInput();
+
+  const [tasks, handleProcess] = useProcessTasks(
+    copyHandler,
+    // @ts-ignore
+    // temporarily ignore error that items is missing
+    // until PR that refactors useProcessTasks is merged
+    fileDataItems,
+    {
+      concurrency: 1,
+    }
+  );
+
+  const destinationListWithoutSlashes =
+    // handle root bucket access grant
+    key === ''
+      ? ['']
+      : // handle subfolder inside root access grant
+      key && current?.prefix == ''
+      ? ['', ...key.split('/').slice(0, -1)]
+      : // regular access that starts at prefix (not root bucket)
+      key.includes('/')
+      ? key.split('/').slice(0, -1)
+      : [];
+
+  const [destinationList, onSetDestinationList] = useState(
+    destinationListWithoutSlashes
+  );
+
+  const taskCounts = getTaskCounts(tasks);
+  const { disableCancel, disableClose, disablePrimary } =
+    getActionViewDisabledButtons(taskCounts);
+
+  const onActionStart = () => {
+    if (!destinationList) return;
+    handleProcess({
+      config: getInput(),
+      prefix: getDestinationListFullPrefix(destinationList),
+    });
+  };
+
+  const onActionCancel = () => {
+    tasks.forEach((task) => {
+      if (isFunction(task.cancel)) task.cancel();
+    });
+  };
+
+  const onExit = () => {
+    // clear files state
+    dispatchStoreAction({ type: 'RESET_LOCATION_ITEMS' });
+    // clear selected action
+    dispatchStoreAction({ type: 'RESET_ACTION_TYPE' });
+    if (isFunction(_onExit)) _onExit();
```

Comment:
nit: can we use optional chaining ? since `onExit?: () => void`
```suggestion
    _onExit?.();
```

Reply:
🍕

Reply:
The preference towards `isFunction` is due to `onExit` being provided by the consumer who may pass a non-function

Reply:
The type should prevent that mistake

Reply:
Not everyone uses typescript